streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -673,44 +772,11 @@ private void handleRemovedTasksFromStateUpdater() {
         final Set<Task> tasksToCloseDirty = new TreeSet<>(Comparator.comparing(Task::id));
 
         for (final Task task : stateUpdater.drainRemovedTasks()) {
-            final TaskId taskId = task.id();
             Set<TopicPartition> inputPartitions;
             if ((inputPartitions = tasks.removePendingTaskToRecycle(task.id())) != null) {
-                try {
-                    final Task newTask = task.isActive() ?
-                        convertActiveToStandby((StreamTask) task, inputPartitions) :
-                        convertStandbyToActive((StandbyTask) task, inputPartitions);
-                    newTask.initializeIfNeeded();
-                    stateUpdater.add(newTask);
-                } catch (final RuntimeException e) {
-                    final String uncleanMessage = String.format("Failed to recycle task %s cleanly. " +
-                        "Attempting to handle remaining tasks before re-throwing:", taskId);
-                    log.error(uncleanMessage, e);
-
-                    if (task.state() != State.CLOSED) {
-                        tasksToCloseDirty.add(task);
-                    }
-
-                    taskExceptions.putIfAbsent(taskId, e);
-                }
+                recycleTask(task, inputPartitions, tasksToCloseDirty, taskExceptions);
             } else if (tasks.removePendingTaskToCloseClean(task.id())) {
-                try {
-                    task.suspend();
-                    task.closeClean();
-                    if (task.isActive()) {
-                        activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(task.id());
-                    }
-                } catch (final RuntimeException e) {
-                    final String uncleanMessage = String.format("Failed to close task %s cleanly. " +
-                        "Attempting to handle remaining tasks before re-throwing:", task.id());
-                    log.error(uncleanMessage, e);
-
-                    if (task.state() != State.CLOSED) {
-                        tasksToCloseDirty.add(task);
-                    }
-
-                    taskExceptions.putIfAbsent(task.id(), e);
-                }
+                closeTaskClean(task, tasksToCloseDirty, taskExceptions);
```

Comment:
And to further remove duplicated logic, I think we can consolidate the `handleRemovedTasksFromStateUpdater` and `handleRestoredTasks` into e.g. a `tryHandleTasksFromStateUpdater` which checks if the task should be 1) recycled, 2) closed-clean / dirty, 3) update input partitions (and then give back to state updater). If the task falls on any of these conditions, return true otherwise false.

Then in `handleRemovedTasksFromStateUpdater`:

```
if (!tryHandleTasksFromStateUpdater(..)) {
    throw new IllegalStateException("this should not happen");
}
```

And in `handleRestoredTasksFromStateUpdater`:

```
if (!tryHandleTasksFromStateUpdater(..)) {
    transitRestoredTaskToRunning(task);
}
```

WDYT?

Reply:
I think reducing code duplication makes a lot of sense. However, can we postpone this refactorings after we have a running system with state updater?
I tried to apply your suggestions regarding `tryHandleTasksFromStateUpdater()` but I do not like the idea of using a boolean to distinguish between removed and restored tasks. I think it would be better to merge the queue of restored and removed tasks as I proposed in our last discussion.

Reply:
Sounds good :) Let's do the queue merging in follow-up PRs then.